### PR TITLE
diffusion: implement the covariance based estimator for diffusion

### PR DESCRIPTION
**Pull request by JoepVanlier**
**Why this PR?**
This is the first part of some of the new diffusion estimation work. This implements the CVE method.

Benefits:
- Provides a really fast unbiased estimator that provides a better alternative to OLS/GLS for cases where the `SNR > 1`.
- Future extension is possible towards methods where we take into account the bias caused by substrate fluctuations (whether this is relevant for us is still T.B.D).
- Future extension to incorporating the static localization uncertainty obtained from Gaussian refinement is possible, this reduces the problem by one parameter and reduces spread of the diffusion estimates.

Below you see a figure where this method was using on a 50 point trace with 20 random points missing. The solid lines with the symbols represent the mean, while the lines indicate `mean +- standard deviation`. You can see that all three methods are unbiased for `SNR > 1`.

image

Note that if you provide the localization uncertainty, the variance goes down (as expected).

image

One thing to keep in mind when dealing with these estimates is that the distribution of the estimates is not very normal. It is skewed towards positive values. While the `mean` is asymptotically correct and unbiased, the `median` will not be at the true value, and the `mean` can be outlier sensitive.

I have also compared the variance estimators that were provided (and implemented in this PR). In the results below we can see the difference between the theoretical variance estimate (variance estimator with the true parameters filled in) vs the variance actually obtained from running many traces (N=1500). Used here are traces of 50 points with 10 points randomly omitted (meaning 40 points effectively):

Left shows unknown localization uncertainty, right shows with localization uncertainty known (currently not exposed through the API):

image

The relevant papers are: [here](https://backend.orbit.dtu.dk/ws/files/90238574/PhysRevE.89.022726.pdf) and [here](https://arxiv.org/pdf/1504.05462.pdf).

Docs build can be found [here](https://lumicks-pylake.readthedocs.io/en/cve/_api/lumicks.pylake.kymotracker.kymotrack.KymoTrack.html#lumicks.pylake.kymotracker.kymotrack.KymoTrack.estimate_diffusion) and [here](https://lumicks-pylake.readthedocs.io/en/cve/tutorial/kymotracking.html#studying-diffusion-processes). They are usable, but still a bit spartan. Reason for this is that I intend to restructure them when the ensemble MSD PR rolls around. Since that one requires some reworking in how the story is told, I didn't want to add too many docs changes to this PR (to avoid double work).

### CODE REVIEW

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 1]
```diff
@@ -447,3 +452,257 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    dt: float,
+    num_points: int,
+    blur_constant: float = 0,
+    avg_frame_steps: float = 1,
+) -> float:
+    """Expected variance of the diffusion estimate obtained with CVE.
+
+    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+    This function is based on eqn 22 from Vestergaard et al [2] adapted for 1D. See also eqn 17 from
+    Vestergaard et al [1].
```
Comment (ab412) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:469]: Does this still align with the error message in `DiffusionEstimate`?  Do we still expect some bias when using cve as well? I'm referring to this line:
```HTML
        Note: while this provides some measure of uncertainty, these estimates are expected to be
        biased and should not be used to weight multiple diffusion estimates!
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 2]
```diff
@@ -447,3 +452,257 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    .. [1] Vestergaard, C. L., Blainey, P. C., & Flyvbjerg, H. (2014). Optimal estimation of
+           diffusion coefficients from single-particle trajectories. Physical Review E, 89(2),
+           022726.
+    """
+    epsilon = variance_loc / (diffusion_constant * dt) - 2.0 * blur_constant
+    numerator = 6.0 * avg_frame_steps**2 + 4.0 * epsilon * avg_frame_steps + 2.0 * epsilon**2
+    denominator = num_points * avg_frame_steps**2
+    term1 = numerator / denominator
+    term2 = (4.0 * (avg_frame_steps + epsilon) ** 2) / (num_points**2 * avg_frame_steps**2)
+    return (term1 + term2) * diffusion_constant**2
```
Comment (ab412) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:500]: I would use built-in functions where applicable (`math.pow()`) - unless there's a reason not to. Perhaps also handle divide by zero scenarios.
```suggestion
    try
        epsilon = variance_loc / (diffusion_constant * dt) - 2.0 * blur_constant
        numerator = 6.0 * pow(avg_frame_steps,2) + 4.0 * epsilon * avg_frame_steps + 2.0 * pow(epsilon, 2)
        denominator = num_points * pow(avg_frame_steps, 2)
        term1 = numerator / denominator
        term2 = (4.0 * pow((avg_frame_steps + epsilon) ,2)) / pow(num_points,2) * pow(avg_frame_steps,2))
        return (term1 + term2) * pow(diffusion_constant,2)
    except ZeroDivisionError:
         raise ValueError("Diffusion constant, blur constant and time step parameters should not be 0")
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 3]
```diff
@@ -447,3 +452,257 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    variance_variance_loc: float,
+    dt: float,
+    num_points: int,
+    blur_constant: float = 0,
+    avg_frame_steps: float = 1,
+) -> float:
+    """Expected variance of the diffusion estimate obtained with CVE.
+
+    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+    This function is based on eqn 24 from Vestergaard et al [2], but adapted for 1D.
```
Comment (ab412) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:515]:
```suggestion
    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
    This function is based on equation 24 from Vestergaard et al [2], but adapted for 1D.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 4]
```diff
@@ -447,3 +452,257 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+        .. math::
+            R = \frac{1}{\Delta t} \int_{0}^{\Delta t}S(t) \left(1 - S(t)\right)dt
+
+        with
+
+        .. math::
+            S(t) = \int_{0}^{t} c(t') dt'
+
+        For no motion blur, we obtain: R = 0, while a continuously open shutter over the exposure
```
Comment (ab412) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:606]:
```suggestion
        When there is no motion blur, we obtain: R = 0, whereas a continuously open shutter over the exposure
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 1 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:469]: Yes. The std error reported is biased. It's the best estimate we got at this point, but it's biased when inserting the estimates we obtained for `D` and `sigma` (I have also confirmed that there is a bias when simulating this). The only time we see no bias is when I compute it with the true values for `D` and `sigma`, but these will not be known when doing a real experiment. The paper also explicitly mentions that they should not be used to weight data. Look for `For importance weighting of estimates...` on page 7.

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 2 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:500]: Why would I prefer `math.pow` over `**`? Personally I find the operator `**` much more readable.

That's a good point about division by zero. Though, the error message in the division by zero exception you suggest is not correct. The blur constant can definitely be zero (and in our case always is). The diffusion constant can actually also be zero numerically, and this indeed suggests that I should rewrite the equation a little bit.

Time step should not be zero; but I don't think it can be by construction.

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 3 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 2 above)
Comment (ab412) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py]: > Why would I prefer math.pow over **? Personally I find the operator ** much more readable and this code is not exactly performance critical.

Readability was the main thing, and also considering operator `**` has differing meanings based on context one might prefer the distinction. Comes down to preference and the intended audience I suppose. 

> Edit: I've also added a test which specifically includes a test with zero diffusion constant.

Even better!

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 1 above)
Comment (ab412) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py]: I see. Then perhaps the wording can be altered to indicate the reasoning behind the note? The idea being providing enough information from the API without needing to bounce back and forth with the source. What do you think of the something like following?

`While this provides some measure of uncertainty, the estimate should be avoided when weighting multiple diffusion estimates.  This is to prevent complications such as bias due to correlations between the estimated parameters and the estimated variances`

Comment (aafkevandenberg) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py]: And maybe add the  suggestion from the author in the same paragraph 'when cal-
culating the weighted mean of estimates from time series
of different lengths—the length N of a time series should
be used as weight when possible, since it is known exactly'

Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py]: In the next PR, we'll be adding some functionality for actually weighting estimates coming from different tracks, so that the user never has to worry about this.

But yeah, I agree with both of these. How about for now we go with
```python
        Note: While this provides some measure of uncertainty, the estimate should not be used for
        weighting multiple diffusion estimates. This is to prevent complications such as bias due
        to correlations between the estimated parameters and estimated variances. Instead, when
        calculating the weighted mean of estimates from time series of different lengths—the length
        N of a time series should be used as weight when possible, since it is known exactly.
```

And then once we have the functionality on the group, we refer to the group functions there (e.g. note that you can compute diffusion constants for a group of tracks in the following way etc etc....).

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 5]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+def _var_cve_unknown_var(
+    diffusion_constant: float,
+    variance_loc: float,
+    dt: float,
+    num_points: int,
+    blur_constant: float = 0,
+    avg_frame_steps: float = 1,
+) -> float:
+    """Expected variance of the diffusion estimate obtained with CVE.
+
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:469]:
```suggestion
    """Expected variance of the diffusion estimate obtained with CVE
         when the localization variance is not known a priori.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 6]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        Time step
+    num_points : int
+        Number of points in the trace
+    blur_constant : float
+        Motion blur constant. See :func:`lumicks.pylake.kymotracker.detail.msd_estimation._cve` for
+        more information.
+    avg_frame_steps : float
+        Average frame steps. This number is the average of the time between two localizations in
+        frames mean(delta time between frame / frame timestep). If all frames had successful
+        localization, this constant will be 1 [2].
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:540]:
```suggestion
        localization, this constant will be 1 [3]_.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 7]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        Time step
+    num_points : int
+        Number of points in the trace
+    blur_constant : float
+        Motion blur constant. See :func:`lumicks.pylake.kymotracker.detail.msd_estimation._cve` for
+        more information.
+    avg_frame_steps : float
+        Average frame steps. This number is the average of the time between two localizations in
+        frames mean(delta time between frame / frame timestep). If all frames had successful
+        localization, this constant will be 1.
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:490]:
```suggestion
    avg_frame_steps : float
        Average frame steps. This number is the average of the time between two localizations in
        frames mean(delta time between frame / frame timestep). If all frames had successful
        localization, this constant will be 1. [1]_
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 8]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+    Note that this estimator is only valid in the case of pure diffusion without drift.
+
+    References
+    ----------
+    .. [1] Vestergaard, C. L., Blainey, P. C., & Flyvbjerg, H. (2014). Optimal estimation of
+           diffusion coefficients from single-particle trajectories. Physical Review E, 89(2),
+           022726.
+    .. [2] Vestergaard, C. L. (2016). Optimizing experimental parameters for tracking of diffusing
+           particles. Physical Review E, 94(2), 022401.
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:583]: this should move to the bottom to be consistent with other docstrings. also watch out for the numbering, needs to be 4 & 5 so the docs would be formatted properly

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 9]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    dt: float,
+    blur_constant: float = 0,
+    variance_loc: Optional[float] = None,
+    variance_variance_loc: Optional[float] = None,
+) -> tuple:
+    """Covariance based estimator.
+
+    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+    This estimator was introduced in the work of Vestergaard et al [1]. The correction for
+    missing data was introduced in [2].
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:573]:
```suggestion
    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
    This estimator was introduced in the work of Vestergaard et al [1]_. The correction for
    missing data was introduced in [2]_.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 10]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        The variance of the localization variance estimate (needed if variance_loc is provided).
+
+    Note
+    ----
+    With the SNR defined as:
+
+    .. math::
+        \sqrt{D \Delta t} / \sigma
+
+    With D the diffusion constant and :math:`\sigma`
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:628]:
```suggestion
    With :math:`D` the diffusion constant and :math:`\sigma`
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 11]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+    # Determine the average step size
+    average_frame_step = np.mean(np.diff(frame_indices))
+    avg_dt = average_frame_step * dt
+
+    dx = np.diff(x)
+    mean_dx_squared = np.mean(dx**2)
+
+    if len(x) < 3:
+        raise RuntimeError("Insufficient intervals for using the CVE")
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:642]: maybe move towards the top and just throw right away?

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 12]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+    if not variance_loc:
+        # Estimate the variance based on this frame
+        mean_dx_consecutive = np.mean(dx[1:] * dx[:-1])
+        # Equation 21 from [2] adapted for 1D.
+        diffusion_constant = mean_dx_squared / (2 * avg_dt) + mean_dx_consecutive / avg_dt
+        # Equation 15 from [1]. Note that static loc uncertainty is not affected by frame skipping.
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:649]: the reference  numbers will have to be changed

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 13]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        variance_loc = (
+            blur_constant * mean_dx_squared + (2 * blur_constant - 1) * mean_dx_consecutive
+        )
+        var_diffusion = _var_cve_unknown_var(
+            diffusion_constant, variance_loc, dt, len(x), blur_constant, average_frame_step
+        )
+    else:
+        # We know the variance in advance. Equation 23 from [2] adapted for 1D.
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:657]: same with the reference number

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 14]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        Time step
+    blur_constant : float
+        Motion blur coefficient. Should be between 0 and 1/4.
+    unit : str
+        Unit that the diffusion constant is specified in.
+    unit_label : str
+        Unit in TeX format used for plotting labels.
+    """
+
+    # We hardcode the blur constant for confocal for now (no motion blur)
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:703]: 👍

File: lumicks/pylake/kymotracker/kymotrack.py [hunk 15]
```diff
@@ -336,6 +336,22 @@ def plot_msd(self, max_lag=None, **kwargs):
...
+
+        1. Covariance based estimator (CVE)
+
+        The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+        This estimator was introduced in the work of Vestergaard et al [5]_. The correction for
+        missing data was introduced in [6]_. The CVE is unbiased and practically optimal when the
+        signal-to-noise ratio (SNR) is bigger than 1. In this context, the SNR is defined
+        as: :math:`\sqrt{D \Delta t} / \sigma`.
+
+        2. MSD-based estimators
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/kymotrack.py:353]: might be good to specifically state which ones these are (even though it's probably obvious)
```suggestion
        2. MSD-based estimators (OLS, GLS)
```

File: lumicks/pylake/kymotracker/tests/test_derived_quantities/test_msd.py [hunk 16]
```diff
@@ -2,8 +2,11 @@
 import contextlib
 from lumicks.pylake.kymotracker.detail.msd_estimation import *
 from lumicks.pylake.kymotracker.detail.msd_estimation import (
+    _var_cve_known_var,
+    _var_cve_unknown_var,
     _msd_diffusion_covariance,
     _diffusion_ols,
+    _cve,
 )
```
Comment (rpauszek) [inline lumicks/pylake/kymotracker/tests/test_derived_quantities/test_msd.py:10]: hmm, why do we have import all and then import specific functions?

**Review by rpauszek — APPROVED** [review]
Nice, good to have an estimator that's not based on MSD. Just left some suggestions for minor formatting changes and a couple where the references got mixed up

File: lumicks/pylake/kymotracker/tests/test_derived_quantities/test_msd.py (hunk 16 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/tests/test_derived_quantities/test_msd.py:10]: Private functions are not imported when importing `*` unless explicitly listed in `__all__`.

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 17]
```diff
@@ -12,7 +14,13 @@ class DiffusionEstimate:
...
         Estimate for the diffusion constant.
     std_err : float
-        Standard error,
+        Standard error.
+
+        Note: While this provides some measure of uncertainty, the estimate should not be used for
+        weighting multiple diffusion estimates. This is to prevent complications such as bias due
+        to correlations between the estimated parameters and estimated variances. Instead, when
+        calculating the weighted mean of estimates from time series of different lengths—the length
+        N of a time series should be used as weight when possible, since it is known exactly.
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:23]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 18]
```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    With the SNR defined as:
+
+    .. math::
+        \sqrt{D \Delta t} / \sigma
+
+    With D the diffusion constant and :math:`\sigma`
+    This method requires SNR > 1. Below SNR = 1 it degrades rapidly.
+    """
+    if not 0 <= blur_constant < 0.25:
+        raise ValueError("Motion blur constant should be between 0 and 1/4")
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:632]: Good catch! Fixing.

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 11 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:642]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 19]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+           022726.
+    """
+    # Note that it uses a different definition of epsilon to circumvent a division by zero for D=0.
+    epsilon = variance_loc / dt - 2.0 * blur_constant * diffusion_constant
+    avg_diff = avg_frame_steps * diffusion_constant
+    numerator = 6.0 * avg_diff**2 + 4.0 * epsilon * avg_diff + 2.0 * epsilon**2
+    denominator = num_points * avg_frame_steps**2
+    term1 = numerator / denominator
+    term2 = 4.0 * (avg_diff + epsilon) ** 2 / (num_points**2 * avg_frame_steps**2)
+    return term1 + term2
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:528]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 20]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    dt: float,
+    blur_constant: float = 0,
+    variance_loc: Optional[float] = None,
+    variance_variance_loc: Optional[float] = None,
+) -> tuple:
+    r"""Covariance based estimator.
+
+    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+    This estimator was introduced in the work of Vestergaard [11]_. The correction for missing
+    data was introduced in [12]_.
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:599]:
```suggestion
    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
    This estimator was introduced in the work of Vestergaard [12]_. The correction for missing
    data was introduced in [13]_.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 21]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+    With :math:`D` the diffusion constant and :math:`\sigma`
+    This method requires SNR > 1. Below SNR = 1 it degrades rapidly [11]_.
+
+    References
+    ----------
+    .. [11] Vestergaard, C. L., Blainey, P. C., & Flyvbjerg, H. (2014). Optimal estimation of
+            diffusion coefficients from single-particle trajectories. Physical Review E, 89(2),
+            022726.
+    .. [12] Vestergaard, C. L. (2016). Optimizing experimental parameters for tracking of diffusing
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:657]:
```suggestion
    .. [12] Vestergaard, C. L., Blainey, P. C., & Flyvbjerg, H. (2014). Optimal estimation of
            diffusion coefficients from single-particle trajectories. Physical Review E, 89(2),
            022726.
    .. [13] Vestergaard, C. L. (2016). Optimizing experimental parameters for tracking of diffusing
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 22]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+    average_frame_step = np.mean(np.diff(frame_indices))
+    avg_dt = average_frame_step * dt
+
+    dx = np.diff(x)
+    mean_dx_squared = np.mean(dx**2)
+
+    if not variance_loc:
+        # Estimate the variance based on this frame
+        mean_dx_consecutive = np.mean(dx[1:] * dx[:-1])
+        # Equation 21 from [12] adapted for 1D.
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:676]:
```suggestion
        # Equation 14 from [12] and 21 from [13] adapted for 1D.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 23]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        diffusion_constant = mean_dx_squared / (2 * avg_dt) + mean_dx_consecutive / avg_dt
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:677]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 24]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        # Equation 15 from [11]. Note that static loc uncertainty is not affected by frame skipping.
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:678]:
```suggestion
        # Equation 15 from [12]. Note that static loc uncertainty is not affected by frame skipping.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 25]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        variance_loc = (
+            blur_constant * mean_dx_squared + (2 * blur_constant - 1) * mean_dx_consecutive
+        )
+        var_diffusion = _var_cve_unknown_var(
+            diffusion_constant, variance_loc, dt, len(x), blur_constant, average_frame_step
+        )
+    else:
+        # We know the variance in advance. Equation 23 from [12] adapted for 1D.
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:686]:
```suggestion
        # We know the variance in advance. Equation 16 from [12] and 23 from [13] adapted for 1D.
```

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 26]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        diffusion_constant = (mean_dx_squared - 2.0 * variance_loc) / (
+            2.0 * (avg_dt - 2.0 * blur_constant * dt)
+        )
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:689]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 27]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+    # Determine the average step size
+    average_frame_step = np.mean(np.diff(frame_indices))
+    avg_dt = average_frame_step * dt
+
+    dx = np.diff(x)
+    mean_dx_squared = np.mean(dx**2)
+
+    if not variance_loc:
+        # Estimate the variance based on this frame
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:674]: Is it just one frame? I thought x (and therefore dx) are determined from many consecutive frames?

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 28]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+        Time step
+    blur_constant : float
+        Motion blur coefficient. Should be between 0 and 1/4.
+    unit : str
+        Unit that the diffusion constant is specified in.
+    unit_label : str
+        Unit in TeX format used for plotting labels.
+    """
+
+    # We hardcode the blur constant for confocal for now (no motion blur)
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/detail/msd_estimation.py:732]: Hm, I would probably move this comment into the calling function of `kymotrack` (see comment below).

File: lumicks/pylake/kymotracker/kymotrack.py [hunk 29]
```diff
@@ -397,6 +421,16 @@ def estimate_diffusion(self, method, max_lag=None):
             )
 
         frame_idx, positions = np.array(self.time_idx, dtype=int), np.array(self.position)
+        unit_labels = {
+            "unit": f"{self._kymo._calibration.unit}^2 / s",
+            "unit_label": f"{self._kymo._calibration.unit_label}$^2$/s",
+        }
+
+        if method == "cve":
+            return estimate_diffusion_cve(
```
Comment (tobiasjj) [inline lumicks/pylake/kymotracker/kymotrack.py:430]:
```suggestion
            # We hardcode the blur constant for confocal for now (no motion blur)
            return estimate_diffusion_cve(
```

**Review by tobiasjj — APPROVED** [review]
Nice work! Only some minor nit picks and suggestions for adding references to the initial publication of Vestergaard from the year 2014 (for symmetry reasons). LGTM!

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 30]
```diff
@@ -38,14 +46,17 @@ def __float__(self):
 
 
 def calculate_msd(frame_idx, position, max_lag):
-    """Estimate the Mean Square Displacement (MSD) for various time lags.
+    r"""Estimate the Mean Square Displacement (MSD) for various time lags.
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:49]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 31]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+def _var_cve_known_var(
+    diffusion_constant: float,
+    variance_loc: float,
+    variance_variance_loc: float,
+    dt: float,
+    num_points: int,
+    blur_constant: float = 0,
+    avg_frame_steps: float = 1,
+) -> float:
+    r"""Expected variance of the diffusion estimate obtained with CVE.
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:540]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 32]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+    This function is based on equation 24 from Vestergaard [10]_, but adapted for 1D.
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:543]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py [hunk 33]
```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
...
+
+            \frac{1}{N} \sum_{i=1}^{N_{frames - 1}} \frac{t_{i+1} - t_{i}}{dt}
+
+        where :math:`dt` is the time step. If all frames had successful localization, this constant
+        will be 1 [10]_.
+
+    References
+    ----------
+    .. [10] Vestergaard, C. L. (2016). Optimizing experimental parameters for tracking of diffusing
+            particles. Physical Review E, 94(2), 022401.
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:574]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 20 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:599]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 21 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:657]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 23 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:677]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 24 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:678]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 22 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:676]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 25 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:686]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 27 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:674]: Good catch! Clearly that should've been `track`. Fixing.

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 28 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:732]: Yeah... hmm... I wonder why I put it here. My guess is that I moved the =0 upstream at some point during writing this and forgot to move the comment with. Anyhow, good catch! 👍

File: lumicks/pylake/kymotracker/kymotrack.py (hunk 29 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/kymotrack.py:430]: 👍

File: lumicks/pylake/kymotracker/detail/msd_estimation.py (hunk 23 above)
Comment (tobiasjj) [reply lumicks/pylake/kymotracker/detail/msd_estimation.py:672]: 👍
